src/yaml-support/yaml-snippet.ts

```diff
@@ -0,0 +1,55 @@
+import * as vscode from 'vscode';
+import * as path from 'path';
+import * as fs from 'fs';
+import * as yaml from 'js-yaml';
+
+/// internal representation of a yaml code snippet corresponding to vscode.CompletionItemProvider
+export interface CodeSnippet {
+    readonly name : string;
+    readonly label : string;
+    readonly description : string;
+    readonly body : string;
+}
+
+
+/**
+ * A kubernetes completion provider provides yaml code snippets for kubernetes, eg: service, deployment.
+ */
+export class KubernetesCompletionProvider implements vscode.CompletionItemProvider {
+    // storing all loaded yaml code snippets from ../../snippets folder
+    private _snippets: CodeSnippet[] = [];
+
+    // default constructor
+    public constructor() {
+        this._loadCodeSnippets();
+    }
+
+    // provide code snippets for vscode
+    public provideCompletionItems(doc: vscode.TextDocument, pos: vscode.Position) {
+        const wordPos = doc.getWordRangeAtPosition(pos);
+        const word = doc.getText(wordPos);
+
+        return this._filterCodeSnippets(word).map((snippet: CodeSnippet): vscode.CompletionItem =>  {
+            const item = new vscode.CompletionItem(snippet.label, vscode.CompletionItemKind.Snippet);
+            item.insertText = new vscode.SnippetString(snippet.body);
+            item.documentation = snippet.description;
+            return item;
+        });
+    }
+
+    // load yaml code snippets from ../../snippets folder
+    private _loadCodeSnippets(): void {
+        const snippetRoot = path.join(__dirname, '../../../snippets');
+        this._snippets  = fs.readdirSync(snippetRoot)
+            .filter((filename: string): boolean => filename.endsWith('-snippet.yaml'))
```

Comment:
why need to filter `-snippet.yaml` files?

Reply:
shall we load arbitrary yaml files under snippets?

Reply:
@andxu might as well! if somebody chooses to add more snippets in the future, they wouldn't need to hunt through this code to identify why their file didn't load.